rust/src/lib.rs

```diff
@@ -310,6 +310,59 @@ pub unsafe extern "C" fn zcashlc_init_data_database(
     unwrap_exc_or(res, -1)
 }
 
+/// A struct that contains a 16-byte account uuid along with key derivation metadata for that
+/// account.
+///
+/// A returned value containing the all-zeros seed fingerprint and/or u32::MAX for the
+/// hd_account_index indicates that no derivation metadata is available.
+#[repr(C)]
+pub struct FfiAccount {
+    uuid_bytes: [u8; 16],
+    account_name: *mut c_char,
+    key_source: *mut c_char,
+    seed_fingerprint: [u8; 32],
+    hd_account_index: u32,
+}
+
+impl FfiAccount {
+    fn from_account(
+        account: &impl zcash_client_backend::data_api::Account<AccountId = AccountUuid>,
+    ) -> Self {
+        let derivation = account.source().key_derivation();
+        FfiAccount {
+            uuid_bytes: account.id().expose_uuid().into_bytes(),
+            account_name: account.name().map_or(ptr::null_mut(), |name| {
+                CString::new(name).unwrap().into_raw()
+            }),
+            key_source: account
+                .source()
+                .key_source()
+                .map_or(ptr::null_mut(), |s| CString::new(s).unwrap().into_raw()),
+            seed_fingerprint: derivation.map_or([0u8; 32], |d| d.seed_fingerprint().to_bytes()),
+            hd_account_index: derivation.map_or(u32::MAX, |d| d.account_index().into()),
+        }
+    }
+}
+
+/// Frees a FfiAccount value
+///
+/// # Safety
+///
+/// - `ptr` must be non-null and must point to a struct having the layout of [`FfiAccount`].
+#[no_mangle]
+pub unsafe extern "C" fn zcashlc_free_account(ptr: *mut FfiAccount) {
+    if !ptr.is_null() {
+        let account: Box<FfiAccount> = unsafe { Box::from_raw(ptr) };
+        if !(account.account_name).is_null() {
+            unsafe { zcashlc_string_free(account.account_name) }
+        }
+        if !(account.key_source).is_null() {
```

Comment:
What is this parens syntax needed for?

Reply:
The compiler was for some reason trying to bind the `!` operator to the pointer, but parenthesizing makes the `is_null` call typecheck.